django/db/models/manager.py

```diff
@@ -242,12 +244,40 @@ class Manager(BaseManager.from_queryset(QuerySet)):
 class ManagerDescriptor(object):
     # This class ensures managers aren't accessible via model instances.
     # For example, Poll.objects works, but poll_obj.objects raises AttributeError.
-    def __init__(self, manager):
+    # It also ensures managers defined on non-abstract base classes are not inherited by child classes.
+    def __init__(self, manager, model):
         self.manager = manager
+        self.model = model
 
     def __get__(self, instance, cls=None):
         if instance is not None:
             raise AttributeError("Manager isn't accessible via %s instances" % cls.__name__)
+
+        if cls._meta.concrete_model == self.model._meta.concrete_model:
+            # When current cls and descriptor's model have the same concrete model
+            # there's nothing to skip, we can use this manager.
+            return self.manager
+
+        # We should skip all concrete classes in cls mro but the first one to ensure
+        # managers defined on non-abstract base classes are not inherited by child
+        # classes. We only use this code when cls is created, before managers are
+        # contributed so it doesn't need to be cached. We also skip two last classes
+        # since it's django.db.models.Model and object.
+        for model in cls.mro()[:-2]:
+            model_is_concrete = not (model._meta.abstract or model._meta.proxy or model._meta.swapped)
```

Comment:
Please do not assume a model only inherits from `django.db.models.Model` I inherit from standard python classes as well. Add check for `issubclass(model, django.db.models.Model)`.